source/custom/directory_features.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package custom
+
+import (
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"sigs.k8s.io/yaml"
+)
+
+const Directory = "/etc/kubernetes/node-feature-discovery/custom.d"
+
+// getDirectoryFeatureConfig returns features configured in the "/etc/kubernetes/node-feature-discovery/custom.d"
+// host directory, which can be populated e.g. by ConfigMaps
+func getDirectoryFeatureConfig() []FeatureSpec {
+
+	features := make([]FeatureSpec, 0)
+
+	log.Printf("DEBUG: getting files in %s", Directory)
+	files, err := ioutil.ReadDir(Directory)
+	if err != nil {
+		if os.IsNotExist(err) {
+			log.Printf("DEBUG: custom config directory %q does not exist", Directory)
+		} else {
+			log.Printf("ERROR: unable to access custom config directory %q, %v", Directory, err)
+		}
+		return features
+	}
+
+	for _, file := range files {
+		configFile := filepath.Join(Directory, file.Name())
+
+		if file.IsDir() {
+			//log.Printf("DEBUG: skipping dir %q", configFile)
+			continue
```

Comment:
Hmm, I started to think that maybe we should recurse into subdirectories as well. At least one level deep. This would make it much more user-friendly to drop in (mount) multiple config files under `custom.d/`. Each configmap would just be mounted in a separate subdir, that is.

I think with the current scheme it is only possible to use multiple configmaps with `subPath` mount which a bit cumbersome (IMO), and more importantly, prevents nfd-worker from getting confimap updates (i.e. the config inside the container is static even if the configmap(s) changed).

Reply:
Oh, good catch! I thought I can mount multiple CMs into the same path as long as the filenames in their data is unique, but that doesn't work. Will fix this and extend the e2e test to use 2 CMs.

Reply:
> I thought I can mount multiple CMs into the same path as long as the filenames in their data is unique, but that doesn't work.

Unfortunately that's not the case 😞  It's possible with the `subPath` directive, but, it has it's drawbacks as explained above.

> Will fix this and extend the e2e test to use 2 CMs.

That would be great 👍